File: SideChain/types/attribute.go
package types

import (
	"errors"
	"io"

	"github.com/elastos/Elastos.ELA/common"
)

type AttributeUsage byte

const (
	Nonce          AttributeUsage = 0x00
	Script         AttributeUsage = 0x20
	Memo           AttributeUsage = 0x81
	Description    AttributeUsage = 0x90
	DescriptionUrl AttributeUsage = 0x91
)

func (u AttributeUsage) Name() string {
	switch u {
	case Nonce:
		return "Nonce"
	case Script:
		return "Script"
	case Memo:
		return "Memo"
	case Description:
		return "Description"
	case DescriptionUrl:
		return "DescriptionUrl"
	default:
		return "Unknown"
	}
}

func IsValidAttributeType(usage AttributeUsage) bool {
	switch usage {
	case Nonce, Script, Memo, Description, DescriptionUrl:
		return true
	}
	return false
}

type Attribute struct {
	Usage AttributeUsage
	Data  []byte
}

func (a Attribute) String() string {
	return "Attribute: {\n\t\t" +
		"Usage: " + a.Usage.Name() + "\n\t\t" +
		"Data: " + common.BytesToHexString(a.Data) + "\n\t\t" +
		"}"
}

func NewAttribute(u AttributeUsage, d []byte) Attribute {
	return Attribute{u, d}
}

func (a *Attribute) Serialize(w io.Writer) error {
	if err := common.WriteUint8(w, byte(a.Usage)); err != nil {
		return errors.New("Transaction attribute Usage serialization error.")
	}
	if !IsValidAttributeType(a.Usage) {
		return errors.New("[Attribute error] Unsupported attribute Description.")
	}
	if err := common.WriteVarBytes(w, a.Data); err != nil {
		return errors.New("Transaction attribute Data serialization error.")
	}
	return nil
}

func (a *Attribute) Deserialize(r io.Reader) error {
	val, err := common.ReadBytes(r, 1)
	if err != nil {
		return errors.New("Transaction attribute Usage deserialization error.")
	}
	a.Usage = AttributeUsage(val[0])
	if !IsValidAttributeType(a.Usage) {
		return errors.New("[Attribute error] Unsupported attribute Description.")
	}
	a.Data, err = common.ReadVarBytes(r, common.MaxVarStringLength,
		"attribute data")
	if err != nil {
		return errors.New("Transaction attribute Data deserialization error.")
	}
	return nil
}

File: SPV.CPP/SDK/Plugin/Transaction/Attribute.cpp
// Copyright (c) 2012-2018 The Elastos Open Source Project
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "Attribute.h"

#include <Common/Log.h>
#include <Common/Utils.h>

namespace Elastos {
	namespace ElaWallet {

		Attribute::Attribute() :
			_usage(Nonce) {
		}

		Attribute::Attribute(const Attribute &attr) {
			operator=(attr);
		}

		Attribute::Attribute(Attribute::Usage usage, const bytes_t &data) :
			_usage(usage),
			_data(data) {

		}

		Attribute::~Attribute() {
		}

		Attribute& Attribute::operator=(const Attribute &attr) {
			this->_usage = attr._usage;
			this->_data = attr._data;
			return *this;
		}

		Attribute::Usage Attribute::GetUsage() const {
			return _usage;
		}

		const bytes_t &Attribute::GetData() const {
			return _data;
		}

		bool Attribute::IsValid() const {
			return (_usage == Attribute::Usage::Description ||
					_usage == Attribute::Usage::DescriptionUrl ||
					_usage == Attribute::Usage::Memo ||
					_usage == Attribute::Usage::Script ||
					_usage == Attribute::Usage::Nonce ||
					_usage == Attribute::Usage::Confirmations);
		}

		size_t Attribute::EstimateSize() const {
			size_t size = 0;
			ByteStream stream;

			size += 1;
			size += stream.WriteVarUint(_data.size());
			size += _data.size();

			return size;
		}

		void Attribute::Serialize(ByteStream &stream) const {
			stream.WriteUint8(_usage);
			stream.WriteVarBytes(_data);
		}

		bool Attribute::Deserialize(const ByteStream &stream) {
			if (!stream.ReadBytes(&_usage, 1)) {
				Log::error("Attribute deserialize usage fail");
				return false;
			}

			if (!IsValid()) {
				Log::error("invalid attribute usage: {}", (uint8_t)_usage);
				return false;
			}

			if (!stream.ReadVarBytes(_data)) {
				Log::error("Attribute deserialize data fail");
				return false;
			}

			return true;
		}

		nlohmann::json Attribute::ToJson() const {
			nlohmann::json j;
			j["Usage"] = _usage;
			j["Data"] = _data.getHex();

			return j;
		}

		void Attribute::FromJson(const nlohmann::json &j) {
			_usage = j["Usage"].get<Usage>();
			_data.setHex(j["Data"].get<std::string>());
		}

		bool Attribute::operator==(const Attribute &a) const {
			return _usage == a._usage && _data == a._data;
		}

		bool Attribute::operator!=(const Attribute &a) const {
			return !operator==(a);
		}

	}
}